### DISCUSSION TITLE
Multiple Validation Sets

### DISCUSSION THREAD

**Discussion by Jose-Bastos** [Lightning-AI/pytorch-lightning#11135] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hello, I'm trying to validate my model on multiple subsets of the initial validation set to compare performance. Reading [this page](https://pytorch-lightning.readthedocs.io/en/stable/guides/data.html) I got the idea that returning a list contaning the multiple Dataloaders would be enough. My val_dataloader method became the following:

![image](https://user-images.githubusercontent.com/73995923/146572624-438bc306-f049-4cba-972a-64162b98294e.png)

But this isn't working properly. I get the following error: _"TypeError: validation_step() takes 3 positional arguments but 4 were given"_
(it worked properly when I only used 1 validation Dataloader).
What am I doing wrong? Can someone help me with this? Or just point me to some more documentation on this.

Thanks in advance!

**Comment by ananthsub** [comment]
Could you share what your validation step implementation is?

**Comment by rohitgr7** [comment, marked as answer]
you must be missing the additional `dataloader_idx` required in the `validation_step` for multiple dataloaders
docs: https://pytorch-lightning.readthedocs.io/en/latest/guides/data.html#multiple-validation-test-predict-dataloaders

**Reply by Jose-Bastos** [reply]
That was it, thanks!

**Reply by Jose-Bastos** [reply]
And maybe you have some docs on how to log each validation subset with a custom name? Right now it seems I can't override the default "dataloader_idx_1,dataloader_idx_2.... etc" names.

Thanks in advance!

**Reply by rohitgr7** [reply]
checkout `add_dataloader_idx` argument here: https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#automatic-logging

**Reply by mariomeissner** [reply]
@rohitgr7 Thank you for your answer, I was looking for the same thing. 

Actually, I want to customize the name given to each entry. Things break in my code when `val/acc` is gone and replaced with `val/acc/dataloader_idx_0`. Ideally, I'd like one validation set to stay as `val/acc` and then assign a special name to the second one, such as `val/acc/secondary`. 

In the `add_dataloader_idx` it says "If False, user needs to give unique names for each dataloader to not mix the values", but how do I do that?

**Reply by rohitgr7** [reply]
@mariomeissner like this:
```py
class LitModel(LightningModule):
    ...
    
    def validation_step(self, batch, batch_idx, dataloader_idx):
        ...
        if dataloader_idx == 0:
            self.log('val/acc', acc, add_dataloader_idx=False)
        elif dataloader_idx == 1:
            self.log('val/acc_secondary', acc, add_dataloader_idx=False)
```

**Reply by mariomeissner** [reply]
@rohitgr7 That makes a lot of sense. I always create new `Accuracy` instances for different metrics, but here we use the same `acc` for both validation datasets. Is that safe?

**Reply by mariomeissner** [reply]
I think it will interfere. For example, I currently calculate `val_acc_best` like this:
```
acc = self.val_acc.compute()
self.val_acc_best.update(acc)
self.log("val/acc_best", self.val_acc_best.compute(), on_epoch=True, prog_bar=True)
```

But this `acc` is mixing both validation datasets now, returning the wrong value, right?

I guess I have to keep separate metrics for each validation set...

**Reply by rohitgr7** [reply]
yes.. you should create a different metric instance for each dataloader as well if you are using Modular metrics. They contain some internal states so mixing up results for different datasets will result in incorrect computation of the final metric.

**Reply by mariomeissner** [reply]
Thanks again for your help!
